### DISCUSSION TITLE
windicss vite integration.

### DISCUSSION THREAD

**Discussion by Manborough** [ElMassimo/vite_ruby#37] [Q&A] [answered]
Does anyone know if you need to do anything further to enable the windicss vite plugin with vite-rails? I am running it in a docker container for development, if that changes anything.

- under app/frontend/entrypoints/application.js I have import 'virtual:windi.css'
- in the vite.config.ts I have imported windicss and inlcuded it as a plugin.
 (it auto generated as a typescript file, though I'm not using it anywhere else)
- it's in my package.json file too.
- And under the application.html.erb I have <%= vite_javascript_tag 'application' %>.

I've set a h1 to have bg-red-300 but it's not working so far.

**Comment by Manborough** [comment]
Looks like the plugin isn't auto looking at .html.erb files.
I assume I can add in some options for the plugin, but I have no idea how to extend the globs list or change the directory to look in app/views.

```
2021-03-17T05:41:42.953Z vite-plugin-windicss:scan:glob globs [ 'index.html', 'src/**/*.{html,vue,md,mdx,pug,jsx,tsx,svelte}' ]
vite_1   | 2021-03-17T05:41:42.982Z vite-plugin-windicss:config {}
```

**Comment by Manborough** [comment]
I've found the documentation for options! I'll test it out in a couple of days when I have time and report back in case anyone else is trying to solve this.

https://github.com/windicss/vite-plugin-windicss/blob/main/packages/plugin-utils/src/options.ts

**Comment by ElMassimo** [comment, marked as answer]
Hi @Manborough!

Check this repo for a working example: https://github.com/ElMassimo/pingcrm-vite

I recently made a PR to Windi CSS to add a `root` option [here](https://github.com/windicss/vite-plugin-windicss/pull/66).

You should specify the `root` to match the Rails root:

```ruby
    WindiCSS({
      root: process.cwd(),
      scan: {
        fileExtensions: ['erb', 'html', 'vue', 'rb', 'jsx', 'tsx'],
        dirs: ['app/views', 'app/helpers', 'app/javascript'],
      },
    }),
``` 

I might add this to a FAQs section in the docs.

**Reply by Manborough** [reply]
Thanks @ElMassimo, I've got it kinda working.
Currently when I start the server it will load the css styles set in a page and display them normally. But if I change anything I'll lose the styling.

If I set them back to the original classes they'll display again. Which would suggest to me that the css file isn't being replaced or regenerated by windi-css. I've been fiddling around with it most of the morning, but haven't made any progress.

Your ReloadOnChange plugin is working fine, so is the stimulus HMR module, so there doesn't seem to be an issue with the vite setup itself.

 Any suggestions?

**Reply by ElMassimo** [reply]
Are you importing `windi.css` in your [entrypoints](https://github.com/ElMassimo/pingcrm-vite/blob/main/app/javascript/entrypoints/application.js#L6)? 

Another thing to try would be to update to the latest version of vite-plugin-windicss, in case there was a bug in the version you are using.

If you can share a minimal repo with the problem, I can take a look.

**Reply by Manborough** [reply]
I tried both ``` windi.css ``` and  ``` 'virtual:windi.css' ``` in my entrypoints folder. Also just tried upgrading the vite-plugin-windicss to the latest.

I shared the repo I'm working out of, there isn't much to it yet as I'm still learning rails.

**Reply by Manborough** [reply]
The order/demo view has a line of styled text as an example

**Reply by ElMassimo** [reply]
In the example repo, `import 'windi.css'` was commented out, but it's needed in order to include the styles.

https://user-images.githubusercontent.com/1158253/111782932-6375c080-8898-11eb-8c7b-022795aa4639.mp4

The docker compose config is missing `npm install` in the Vite image.

If you were using docker, it's likely that when you tried adding the import, the Vite process had not started correctly because the packages were not installed, so `bin/vite dev` does not actually start.

Try running `docker-compose run --rm --entrypoint "npm install" vite` and restarting the process:

```
vite_1   |   vite v2.1.0 dev server running at:
vite_1   |
vite_1   |   > Local:    http://localhost:3036/vite-dev/
vite_1   |   > Network:  http://172.19.0.3:3036/vite-dev/
vite_1   |
vite_1   |   ready in 1857ms.
```

**Reply by Manborough** [reply]
Hmmm I couldn't sworn I had un-commented it, as for the the missing npm install I didn't think that would matter as the dependencies are cached in a volume. Either way I tried out your tips and I'm still getting the same bug.

I downloaded your docker vite example here, https://github.com/ElMassimo/vite_rails_docker_example, to see if I could replicate it outside of my setup.

I bundle installed and npm installed. I added windicsss to the vite-config (+ other config) and ran docker-compose and a separate vite process.

Here's the result (excuse the slightly random cursor pointing)

https://user-images.githubusercontent.com/23373965/111857263-74036500-8984-11eb-90ce-574dd26e1ea3.mov

I'm assuming here that windi-css-vite should pickup on the changed value from bg-red-300 to bg-red-400 and reload the css.

P.S I don't want to be wasting your time with support questions, so let me know if you've had enough of my replies :) I can always install tailwind the old way.

**Reply by ElMassimo** [reply]
> I'm assuming here that windi-css-vite should pickup on the changed value from bg-red-300 to bg-red-400 and reload the css.

That is true for CSS classes used in files that are handled by Vite such as stylesheets, or Vue and React components.

Going back to your previous comment, it's only now I understand what's going on, sorry.

> If I set them back to the original classes they'll display again.

At the moment, the scanning performed by <kbd>vite-plugin-windicss</kbd> is only performed on startup, so this workflow is not yet supported for files that are not handled by Vite, like server-rendered templates.

I'm thinking about adding support for that in `vite-plugin-windicss` or in `vite-plugin-full-reload`.

<hr/>

Starting in `0.2.0`, `vite-plugin-full-reload` will use Vite's file watcher, so this works out of the box.
Planning to do the same for `vite-plugin-windicss` next week.

**Reply by ElMassimo** [reply]
This is now supported in `vite-plugin-windicss` as well: https://github.com/windicss/vite-plugin-windicss/pull/99.

To get the benefit of fast feedback, it's still better to use it in combination with `vite-plugin-full-reload` if adding classes to server templates.

**Reply by Manborough** [reply]
Damn that was fast!

I'll give it a whirl today and see how it goes. Thanks for all your hard work on this @ElMassimo
